Decomposition for adaptive_avg_pool2d

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #84062

This was already implemented as a lowering in https://github.com/pytorch/torchdynamo/pull/962. I'm putting the idea up here ~(I haven't even run this code, so it surely has *many* issues, but I reckon the general idea should hopefully be alright).~ The tests now pass and I corrected the issues that the first implementation had.

lezcano committed:
Decomposition for adaptive_avg_pool2d

This was already implemented as a lowering in https://github.com/pytorch/torchdynamo/pull/962. I'm putting the idea up here (I haven't even run this code, so it surely has *many* issues, but I reckon the general idea should hopefully be alright).

[ghstack-poisoned]

lezcano committed:
Update on "Decomposition for adaptive_avg_pool2d"


This was already implemented as a lowering in https://github.com/pytorch/torchdynamo/pull/962. I'm putting the idea up here (I haven't even run this code, so it surely has *many* issues, but I reckon the general idea should hopefully be alright).

[ghstack-poisoned]

jansel commented on torch/_decomp/decompositions.py:
```diff
@@ -1225,6 +1225,83 @@ def cudnn_batch_norm_backward(
+        return torch.nn.functional.avg_pool2d(input, kernel, stride)
+
+    def start_index(a, b, c):
+        return (a * c).float().div(b).floor().long()
```
converting to float could introduce rounding issues.

Lezcano replied:
That's a good point in abstract, but I doubt it is an issue in practice, as for that to happen you'd need to have a dimension with size 16M, but fair enough.
This is translated from the implementation in core, so we should probably fix it there as well.

jansel commented on torch/_decomp/decompositions.py:
```diff
@@ -1225,6 +1225,83 @@ def cudnn_batch_norm_backward(
+        # is data-dependent. I don't think there's a way to do this that's not data-dependent really
+        maxlength = int(torch.max(length).item())
+        ones = torch.ones(maxlength - 1, dtype=i0.dtype, device=device)
+        idx = torch.cat([i0.unsqueeze(1), ones]).cumsum(dim=1)  # [size, max]
```
torch.cat will result in horrible performance in backends, don't use it in decomps

Lezcano replied:
It is not clear to me how to implement the batched version of `arange` otherwise. What do you propose?

jansel replied:
Could we keep these as separate tensors and not concat them?

Lezcano replied:
ooops. That is certainly the right way of doing this!

jansel commented on torch/_decomp/decompositions.py:
```diff
@@ -1225,6 +1225,83 @@ def cudnn_batch_norm_backward(
+        # Select elements
+        # Doing this logic with arange is intrinsically data-dependent, as the length of the range
+        # is data-dependent. I don't think there's a way to do this that's not data-dependent really
+        maxlength = int(torch.max(length).item())
```
I don't believe this we can trace this.

lezcano committed:
Update on "Decomposition for adaptive_avg_pool2d"


This was already implemented as a lowering in https://github.com/pytorch/torchdynamo/pull/962. I'm putting the idea up here (I haven't even run this code, so it surely has *many* issues, but I reckon the general idea should hopefully be alright).

[ghstack-poisoned]

Lezcano commented:
@jansel Now I did a more serious attempt. Again, not tested, but I got rid of all the smelly points of the code via a reasonable amount of maths.

lezcano committed:
Update on "Decomposition for adaptive_avg_pool2d"


This was already implemented as a lowering in https://github.com/pytorch/torchdynamo/pull/962. I'm putting the idea up here (I haven't even run this code, so it surely has *many* issues, but I reckon the general idea should hopefully be alright).

[ghstack-poisoned]

lezcano committed:
Update on "Decomposition for adaptive_avg_pool2d"


This was already implemented as a lowering in https://github.com/pytorch/torchdynamo/pull/962. I'm putting the idea up here (I haven't even run this code, so it surely has *many* issues, but I reckon the general idea should hopefully be alright).

[ghstack-poisoned]

lezcano committed:
Update on "Decomposition for adaptive_avg_pool2d"


This was already implemented as a lowering in https://github.com/pytorch/torchdynamo/pull/962. I'm putting the idea up here (I haven't even run this code, so it surely has *many* issues, but I reckon the general idea should hopefully be alright).

[ghstack-poisoned]

Chillee commented:
There should be tests for this in `test/test_decomp.py`, I believe.

lezcano committed:
Update on "Decomposition for adaptive_avg_pool2d"


This was already implemented as a lowering in https://github.com/pytorch/torchdynamo/pull/962. I'm putting the idea up here (I haven't even run this code, so it surely has *many* issues, but I reckon the general idea should hopefully be alright).

[ghstack-poisoned]

lezcano committed:
Update on "Decomposition for adaptive_avg_pool2d"


This was already implemented as a lowering in https://github.com/pytorch/torchdynamo/pull/962. I'm putting the idea up here ~(I haven't even run this code, so it surely has *many* issues, but I reckon the general idea should hopefully be alright).~ The tests now pass and I corrected the issues that the first implementation had.

[ghstack-poisoned]

Lezcano commented on torch/_decomp/decompositions.py:
```diff
@@ -1225,6 +1225,100 @@ def cudnn_batch_norm_backward(
+
+        orange = torch.arange(out_size, device=device)
+        i0 = start_index(orange, out_size, in_size)
+        # Let length = end_index - start_index, i.e. the length of the pooling kernels
+        # length.max() can be computed analytically as follows:
+        maxlength = in_size // out_size + 1
+        in_size_mod = in_size % out_size
+        # adaptive = True iff there are kernels with different lengths
+        adaptive = not (in_size_mod == 0 or out_size % in_size_mod == 0)
+        if adaptive:
+            maxlength += 1
+        elif in_size_mod == 0:
+            maxlength -= 1
```
Deducing this is non-trivial. I have a sketch of a proof that goes as follows:

Denote `in_size` by `x` and `out_size` by `y`. Let `f(n) = floor(n * x / y)` and let `g(n) = ceil(n * x / y)`. We want to compute the maximum of the function `h(n) = g(n+1) - f(n)` for `n = 0, ..., y - 1`. 
By dividing with reminder `x` into `y`, that is, writing `x = k*y + s`, we see that `max_n h(n) = k + max_n g_s(n+1) - f_s(n)` where `g_s(n) = ceil(n * s / y)`, `f_s(n)=floor(n * s / y)`. In other words, we can substitute `x` by `x % y`.
This already gives the `maxlength` value for `s = 0`, that is, when `y` divides `x`. This is the case `in_size_mod == 0` in the code.
The case `s = 1` is also direct, and you see that it's also non-adaptive (the max is constant). This falls in the `out_size % in_size_mod == 0` case.
The general case `out_size % in_size_mod == 0` follows by noting that the output is periodic with period `in_size_mod` in this case.
Then you need to show that in the rest of the cases you indeed get two different values. 
I don't have time now to finish the argument, but it should not be too difficult to finish it from here

Lezcano replied:
fwiw, my number theory has always been quite weak, so there might be a much more direct argument looking at generators for multiplicative groups of integers or smth

Lezcano commented:
The errors come from doing the averages in two sums and one division. I'll fix this next week if @jansel believes that this PR would be of interest.

jansel commented:
I don't think this will be performance competitive with the [existing TorchInductor lowering](https://github.com/pytorch/torchdynamo/blob/058b3581bde241ed72b4092d92e561dd9d82fff0/torchinductor/lowering.py#L2237) because of the need to clamp indices and discard loaded data instead of using hardware load mask registers, which aren't yet possible in primtorch.

I would prioritize others ops (still many with neither a decomp nor a lowering!), though this will still be useful for other non-inductor uses of primtorch.

Chillee commented:
I would still be interested in seeing actual performance numbers 🤔

Lezcano commented:
I'll do the following:
- Fix the CI (should be easy) and we can have it merged, as it may be useful for other backends, or to implement the ref in terms of this one
- Next week, I'll benchmark it against the lowering, to see do they fare.

Lezcano commented:
For what is worth, the use of clamp is not too bad, as it discards very few elements (about 5% or less).

Lezcano commented:
Also, I believe that the maths from this PR to compute the max length of the kernel can be used to implement a better lowering, as it would reduce (a bit) the amount of memory that it would avoid having to use the masked loading at all in some cases.

lezcano committed:
Update on "Decomposition for adaptive_avg_pool2d"


This was already implemented as a lowering in https://github.com/pytorch/torchdynamo/pull/962. I'm putting the idea up here ~(I haven't even run this code, so it surely has *many* issues, but I reckon the general idea should hopefully be alright).~ The tests now pass and I corrected the issues that the first implementation had.

[ghstack-poisoned]

Lezcano commented:
@pytorchbot merge -l

lezcano committed:
Update on "Decomposition for adaptive_avg_pool2d"


This was already implemented as a lowering in https://github.com/pytorch/torchdynamo/pull/962. I'm putting the idea up here ~(I haven't even run this code, so it surely has *many* issues, but I reckon the general idea should hopefully be alright).~ The tests now pass and I corrected the issues that the first implementation had.

[ghstack-poisoned]

Lezcano commented:
@pytorchbot merge -l

Lezcano commented:
I benchmarked the lowering against the decomposition, and the decomposition seems to do as good as the lowering really.

@Chillee @jansel This is the first time I benchmark these things, could you please audit the script to make sure that I didn't miss anything?


<details>
<summary>Script</summary>

```python
import torch
from torch.fx.experimental.proxy_tensor import make_fx
from torch.utils.benchmark import Timer, Compare
from torchinductor.compile_fx import compile_fx_inner, cudagraphify
from torchinductor.decomposition import decompositions
from itertools import product
from functools import partial

benchmark_name = "adaptive_avg_pool2d"
Hs = [128, 256, 512]
Ws = [128, 256, 512]
Bs = [128, 256, 512]


def gen_inputs():
    make_arg = partial(torch.randn, dtype=torch.float32, device="cuda")
    for b, h, w in product(Bs, Hs, Ws):
        for h_dif, w_dif in product((-1, 0, 1), repeat=2):
            out_shape = (h // 2 + h_dif, w // 2 + w_dif)
            yield make_arg((b, h, w)), out_shape


def benchmark(label, f, x, out_shape):
    return Timer("f(x)",
                 globals={"f": f, "x": x},
                 label=benchmark_name,
                 description=label,
                 sub_label=f"{tuple(x.shape)} -> {out_shape}",
                 num_threads=torch.get_num_threads()).blocked_autorange()


def compare(x, out_shape):
    def f(x):
        val = torch.ops.aten.adaptive_avg_pool2d(x, out_shape)
        return (val,)

    print(f"{tuple(x.shape)} -> {out_shape}")
    inp = [x]

    decomposed = make_fx(f, decomposition_table=decompositions, tracing_mode="fake")(*inp)
    compiled_decomposed = compile_fx_inner(decomposed, inp)
    yield benchmark("Decomposed", compiled_decomposed, x, out_shape)

    non_decomposed = make_fx(f, tracing_mode="fake")(*inp)
    compiled_nondecomposed = compile_fx_inner(non_decomposed, inp)
    yield benchmark("Lowering", compiled_nondecomposed, x, out_shape)

    cuda_f = cudagraphify(f, inp)
    yield benchmark("Eager", cuda_f, x, out_shape)


results = []
for args in gen_inputs():
    for res in compare(*args):
        results.append(res)

compare = Compare(results)
compare.trim_significant_figures()
compare.print()
```

</details>

<details>
<summary>Results</summary>

```
[---------------------- adaptive_avg_pool2d ----------------------------]
                               |  Decomposed  |  Lowering  |  Eager
32 threads: -------------------------------------------------------------
(128, 128, 128) -> (63, 63)    |      100     |     110    |    137
(128, 128, 128) -> (63, 64)    |      102     |     102    |    121
(128, 128, 128) -> (63, 65)    |      102     |     103    |    133
(128, 128, 128) -> (64, 63)    |      101     |     101    |    130
(128, 128, 128) -> (64, 64)    |       99     |      99    |    117
(128, 128, 128) -> (64, 65)    |       99     |      99    |    123
(128, 128, 128) -> (65, 63)    |      100     |     100    |    100
(128, 128, 128) -> (65, 64)    |      101     |     101    |    119
(128, 128, 128) -> (65, 65)    |      100     |     101    |    133
(128, 128, 256) -> (63, 127)   |      218     |     218    |    281
(128, 128, 256) -> (63, 128)   |      194     |     194    |    240
(128, 128, 256) -> (63, 129)   |      207     |     208    |    247
(128, 128, 256) -> (64, 127)   |      193     |     193    |    259
(128, 128, 256) -> (64, 128)   |      191     |     192    |    230
(128, 128, 256) -> (64, 129)   |      192     |     192    |    233
(128, 128, 256) -> (65, 127)   |      212     |     213    |    280
(128, 128, 256) -> (65, 128)   |      193     |     192    |    236
(128, 128, 256) -> (65, 129)   |      203     |     203    |    247
(128, 128, 512) -> (63, 255)   |      447     |     447    |    570
(128, 128, 512) -> (63, 256)   |      378     |     378    |    471
(128, 128, 512) -> (63, 257)   |      425     |     424    |    487
(128, 128, 512) -> (64, 255)   |      378     |     378    |    518
(128, 128, 512) -> (64, 256)   |      376     |     376    |    456
(128, 128, 512) -> (64, 257)   |      378     |     379    |    467
(128, 128, 512) -> (65, 255)   |      430     |     430    |    568
(128, 128, 512) -> (65, 256)   |      377     |     377    |    468
(128, 128, 512) -> (65, 257)   |      412     |     412    |    482
(128, 256, 128) -> (127, 63)   |      206     |     206    |    265
(128, 256, 128) -> (127, 64)   |      193     |     193    |    235
(128, 256, 128) -> (127, 65)   |      196     |     196    |    259
(128, 256, 128) -> (128, 63)   |      193     |     193    |    253
(128, 256, 128) -> (128, 64)   |      191     |     191    |    230
(128, 256, 128) -> (128, 65)   |      191     |     191    |    241
(128, 256, 128) -> (129, 63)   |      199     |     199    |    265
(128, 256, 128) -> (129, 64)   |      193     |     193    |    234
(128, 256, 128) -> (129, 65)   |      193     |     194    |    258
(128, 256, 256) -> (127, 127)  |      432     |     432    |    553
(128, 256, 256) -> (127, 128)  |      378     |     379    |    470
(128, 256, 256) -> (127, 129)  |      411     |     411    |    487
(128, 256, 256) -> (128, 127)  |      377     |     378    |    514
(128, 256, 256) -> (128, 128)  |      376     |     376    |    445
(128, 256, 256) -> (128, 129)  |      377     |     377    |    459
(128, 256, 256) -> (129, 127)  |      418     |     418    |    546
(128, 256, 256) -> (129, 128)  |      377     |     377    |    470
(128, 256, 256) -> (129, 129)  |      399     |     399    |    487
(128, 256, 512) -> (127, 255)  |      890     |     890    |   1120
(128, 256, 512) -> (127, 256)  |      747     |     747    |    918
(128, 256, 512) -> (127, 257)  |      842     |     843    |    949
(128, 512, 128) -> (255, 63)   |      404     |     404    |    524
(128, 512, 128) -> (255, 64)   |      379     |     378    |    463
(128, 512, 128) -> (255, 65)   |      384     |     383    |    509
(128, 512, 128) -> (256, 63)   |      378     |     378    |    502
(128, 512, 128) -> (256, 64)   |      376     |     376    |    449
(128, 512, 128) -> (256, 65)   |      376     |     376    |    478
(128, 512, 128) -> (257, 63)   |      394     |     393    |    526
(128, 512, 128) -> (257, 64)   |      377     |     377    |    468
(128, 512, 128) -> (257, 65)   |      379     |     380    |    510
(128, 512, 256) -> (255, 127)  |      863     |     860    |   1090
(128, 512, 256) -> (255, 128)  |      746     |     747    |    935
(128, 512, 256) -> (255, 129)  |      814     |     814    |    969
(128, 512, 256) -> (256, 127)  |      747     |     746    |   1020
(128, 512, 256) -> (256, 128)  |      746     |     746    |    883
(128, 512, 256) -> (256, 129)  |      747     |     746    |    911
(128, 512, 256) -> (257, 127)  |      832     |     831    |   1090
(128, 512, 256) -> (257, 128)  |      746     |     746    |    936
(128, 512, 256) -> (257, 129)  |      789     |     789    |    967
(128, 512, 512) -> (255, 255)  |     1772     |    1772    |   2236
(128, 512, 512) -> (255, 256)  |     1484     |    1484    |   1820
(128, 512, 512) -> (255, 257)  |     1680     |    1680    |   1880
(128, 512, 512) -> (256, 255)  |     1485     |    1485    |   2044
(128, 512, 512) -> (256, 256)  |     1485     |    1484    |   1770
(128, 512, 512) -> (256, 257)  |     1490     |    1490    |   1800
(128, 512, 512) -> (257, 255)  |     1690     |    1690    |   2222
(128, 512, 512) -> (257, 256)  |     1484     |    1484    |   1815
(128, 512, 512) -> (257, 257)  |     1620     |    1626    |   1880
(256, 128, 128) -> (63, 63)    |      205     |     205    |    257
(256, 128, 128) -> (63, 64)    |      194     |     194    |    215
(256, 128, 128) -> (63, 65)    |      196     |     196    |    235
(256, 128, 128) -> (64, 63)    |      192     |     193    |    231
(256, 128, 128) -> (64, 64)    |      191     |     191    |    213
(256, 128, 128) -> (64, 65)    |      191     |     191    |    220
(256, 128, 128) -> (65, 63)    |      200     |     200    |    244
(256, 128, 128) -> (65, 64)    |      193     |     205    |    225
(256, 128, 128) -> (65, 65)    |      194     |     194    |    234
(256, 128, 256) -> (63, 127)   |      431     |     432    |    496
(256, 128, 256) -> (63, 128)   |      378     |     378    |    427
(256, 128, 256) -> (63, 129)   |      410     |     411    |    440
(256, 128, 256) -> (64, 127)   |      377     |     377    |    468
(256, 128, 256) -> (64, 128)   |      376     |     376    |    422
(256, 128, 256) -> (64, 129)   |      377     |     377    |    421
(256, 128, 256) -> (65, 127)   |      421     |     421    |    495
(256, 128, 256) -> (65, 128)   |      398     |     377    |    424
(256, 128, 256) -> (65, 129)   |      400     |     400    |    439
(256, 128, 512) -> (63, 255)   |      890     |     890    |   1020
(256, 128, 512) -> (63, 256)   |      747     |     747    |    851
(256, 128, 512) -> (63, 257)   |      844     |     844    |    870
(256, 128, 512) -> (64, 255)   |      747     |     747    |    927
(256, 128, 512) -> (64, 256)   |      745     |     745    |    838
(256, 128, 512) -> (64, 257)   |      749     |     749    |    847
(256, 128, 512) -> (65, 255)   |      857     |     856    |   1020
(256, 128, 512) -> (65, 256)   |      747     |     747    |    848
(256, 128, 512) -> (65, 257)   |      820     |     820    |    866
(256, 256, 128) -> (127, 63)   |      404     |     404    |    473
(256, 256, 128) -> (127, 64)   |      378     |     378    |    422
(256, 256, 128) -> (127, 65)   |      384     |     384    |    461
(256, 256, 128) -> (128, 63)   |      378     |     378    |    457
(256, 256, 128) -> (128, 64)   |      376     |     376    |    419
(256, 256, 128) -> (128, 65)   |      394     |     376    |    433
(256, 256, 128) -> (129, 63)   |      393     |     394    |    475
(256, 256, 128) -> (129, 64)   |      377     |     377    |    423
(256, 256, 128) -> (129, 65)   |      380     |     380    |    460
(256, 256, 256) -> (127, 127)  |      859     |     860    |    982
(256, 256, 256) -> (127, 128)  |      746     |     747    |    846
(256, 256, 256) -> (127, 129)  |      816     |     816    |    878
(256, 256, 256) -> (128, 127)  |      746     |     746    |    925
(256, 256, 256) -> (128, 128)  |      746     |     745    |    826
(256, 256, 256) -> (128, 129)  |      746     |     746    |    831
(256, 256, 256) -> (129, 127)  |      830     |     830    |    976
(256, 256, 256) -> (129, 128)  |      746     |     746    |    846
(256, 256, 256) -> (129, 129)  |      792     |     792    |    874
(256, 256, 512) -> (127, 255)  |     1800     |    1780    |   2044
(256, 256, 512) -> (127, 256)  |     1484     |    1484    |   1670
(256, 256, 512) -> (127, 257)  |     1683     |    1680    |   1720
(256, 256, 512) -> (128, 255)  |     1600     |    1485    |   1850
(256, 256, 512) -> (128, 256)  |     1480     |    1484    |   1640
(256, 256, 512) -> (128, 257)  |     1490     |    1489    |   1660
(256, 256, 512) -> (129, 255)  |     1703     |    1705    |   2030
(256, 256, 512) -> (129, 256)  |     1484     |    1484    |   1670
(256, 256, 512) -> (129, 257)  |     1630     |    1630    |   1715
(256, 512, 128) -> (255, 63)   |      799     |     799    |    934
(256, 512, 128) -> (255, 64)   |      746     |     747    |    834
(256, 512, 128) -> (255, 65)   |      757     |     757    |    908
(256, 512, 128) -> (256, 63)   |      747     |     747    |    902
(256, 512, 128) -> (257, 64)   |      747     |     746    |    837
(256, 512, 128) -> (257, 65)   |      750     |     750    |    908
(256, 512, 256) -> (255, 127)  |     1712     |    1712    |   1950
(256, 512, 256) -> (255, 128)  |     1485     |    1485    |   1690
(256, 512, 256) -> (255, 129)  |     1620     |    1621    |   1740
(256, 512, 256) -> (256, 127)  |     1484     |    1484    |   1870
(256, 512, 256) -> (256, 128)  |     1480     |    1485    |   1630
(256, 512, 256) -> (256, 129)  |     1486     |    1486    |   1660
(256, 512, 256) -> (257, 127)  |     1660     |    1660    |   1958
(256, 512, 256) -> (257, 128)  |     1485     |    1484    |   1693
(256, 512, 256) -> (257, 129)  |     1571     |    1571    |   1738
(256, 512, 512) -> (255, 255)  |     3537     |    3658    |   4252
(256, 512, 512) -> (255, 256)  |     2958     |    2959    |   3332
(256, 512, 512) -> (255, 257)  |     3358     |    3361    |   3435
(256, 512, 512) -> (256, 255)  |     2961     |    2961    |   3724
(256, 512, 512) -> (256, 256)  |     2963     |    2963    |   3271
(256, 512, 512) -> (256, 257)  |     2966     |    2966    |   3310
(256, 512, 512) -> (257, 255)  |     3372     |    3375    |   4030
(256, 512, 512) -> (257, 256)  |     3049     |    3089    |   3322
(256, 512, 512) -> (257, 257)  |     3244     |    3243    |   3409
(512, 128, 128) -> (63, 63)    |      403     |     403    |    452
(512, 128, 128) -> (63, 64)    |      378     |     378    |    404
(512, 128, 128) -> (63, 65)    |      383     |     383    |    444
(512, 128, 128) -> (64, 63)    |      378     |     377    |    439
(512, 128, 128) -> (64, 64)    |      376     |     376    |    402
(512, 128, 128) -> (64, 65)    |      376     |     376    |    417
(512, 128, 128) -> (65, 63)    |      396     |     395    |    454
(512, 128, 128) -> (65, 64)    |      378     |     378    |    404
(512, 128, 128) -> (65, 65)    |      394     |     380    |    445
(512, 128, 256) -> (63, 127)   |      920     |     858    |    930
(512, 128, 256) -> (63, 128)   |      747     |     746    |    804
(512, 128, 256) -> (63, 129)   |      813     |     813    |    834
(512, 128, 256) -> (64, 127)   |      747     |     746    |    882
(512, 128, 256) -> (64, 128)   |      746     |     746    |    805
(512, 128, 256) -> (64, 129)   |      746     |     746    |    807
(512, 128, 256) -> (65, 127)   |      834     |     834    |    926
(512, 128, 256) -> (65, 128)   |      746     |     746    |    804
(512, 128, 256) -> (65, 129)   |      792     |     793    |    833
(512, 128, 512) -> (63, 255)   |     1773     |    1770    |   1930
(512, 128, 512) -> (63, 256)   |     1483     |    1484    |   1630
(512, 128, 512) -> (63, 257)   |     1681     |    1682    |   1673
(512, 128, 512) -> (64, 255)   |     1485     |    1485    |   1755
(512, 128, 512) -> (64, 256)   |     1484     |    1484    |   1600
(512, 128, 512) -> (64, 257)   |     1490     |    1490    |   1626
(512, 128, 512) -> (65, 255)   |     1705     |    1710    |   1920
(512, 128, 512) -> (65, 256)   |     1485     |    1485    |   2000
(512, 128, 512) -> (65, 257)   |     2000     |    1632    |   1670
(512, 256, 128) -> (127, 63)   |      900     |     800    |    895
(512, 256, 128) -> (127, 64)   |      746     |     747    |    801
(512, 256, 128) -> (127, 65)   |      756     |     756    |    875
(512, 256, 128) -> (128, 63)   |      747     |     800    |    900
(512, 256, 128) -> (128, 64)   |      800     |     750    |    800
(512, 256, 128) -> (128, 65)   |      746     |     746    |    824
(512, 256, 128) -> (129, 63)   |      777     |     778    |    890
(512, 256, 128) -> (129, 64)   |      747     |     746    |    804
(512, 256, 128) -> (129, 65)   |      751     |     751    |    876
(512, 256, 256) -> (127, 127)  |     1710     |    1800    |   1840
(512, 256, 256) -> (127, 128)  |     1480     |    2000    |   1602
(512, 256, 256) -> (127, 129)  |     1620     |    1618    |   1660
(512, 256, 256) -> (128, 127)  |     1484     |    1484    |   1757
(512, 256, 256) -> (128, 128)  |     1485     |    1485    |   1582
(512, 256, 256) -> (128, 129)  |     1485     |    1486    |   1593
(512, 256, 256) -> (129, 127)  |     1647     |    1648    |   1840
(512, 256, 256) -> (129, 128)  |     1484     |    1484    |   1605
(512, 256, 256) -> (129, 129)  |     1572     |    1570    |   1660
(512, 256, 512) -> (127, 255)  |     3538     |    3539    |   3857
(512, 256, 512) -> (127, 256)  |     2958     |    3095    |   3382
(512, 256, 512) -> (127, 257)  |     3349     |    3377    |   3320
(512, 256, 512) -> (128, 255)  |     2960     |    2961    |   3512
(512, 256, 512) -> (128, 256)  |     2963     |    2963    |   3153
(512, 256, 512) -> (128, 257)  |     2966     |    2966    |   3316
(512, 256, 512) -> (129, 255)  |     3392     |    3403    |   4105
(512, 256, 512) -> (129, 256)  |     2967     |    2960    |   3201
(512, 256, 512) -> (129, 257)  |     3255     |    3255    |   3287
(512, 512, 128) -> (255, 63)   |     1600     |    1590    |   1772
(512, 512, 128) -> (255, 64)   |     1485     |    1480    |   1592
(512, 512, 128) -> (255, 65)   |     1510     |    1510    |   1743
(512, 512, 128) -> (256, 63)   |     1485     |    1480    |   1718
(512, 512, 128) -> (256, 64)   |     1484     |    1484    |   1577
(512, 512, 128) -> (256, 65)   |     1484     |    1484    |   1645
(512, 512, 128) -> (257, 63)   |     1560     |    1560    |   1770
(512, 512, 128) -> (257, 64)   |     1484     |    1550    |   2000
(512, 512, 128) -> (257, 65)   |     1495     |    1495    |   1900
(512, 512, 256) -> (255, 127)  |     3444     |    3429    |   3684
(512, 512, 256) -> (255, 128)  |     2957     |    2958    |   3214
(512, 512, 256) -> (255, 129)  |     3244     |    3244    |   3315
(512, 512, 256) -> (256, 127)  |     2959     |    2959    |   3565
(512, 512, 256) -> (256, 128)  |     2962     |    2962    |   3141
(512, 512, 256) -> (256, 129)  |     2963     |    2963    |   3176
(512, 512, 256) -> (257, 127)  |     3322     |    3322    |   3698
(512, 512, 256) -> (257, 128)  |     2960     |    2960    |   3217
(512, 512, 256) -> (257, 129)  |     3142     |    3140    |   3314
(512, 512, 512) -> (255, 255)  |     7100     |    7100    |   7800
(512, 512, 512) -> (255, 256)  |     5909     |    5910    |   6432
(512, 512, 512) -> (255, 257)  |     6733     |    6730    |   6615
(512, 512, 512) -> (256, 255)  |     5913     |    5913    |   7120
(512, 512, 512) -> (256, 256)  |     5919     |    5920    |   6297
(512, 512, 512) -> (256, 257)  |     5930     |    5921    |   6370
(512, 512, 512) -> (257, 255)  |     6759     |    6760    |   7660
(512, 512, 512) -> (257, 256)  |     5914     |    5913    |   6378
(512, 512, 512) -> (257, 257)  |     6499     |    6500    |   6550
```

</details>

Chillee commented:
Benchmarking looks fine.

@jansel @ngimel Should we delete the lowering in favor of the decomposition then?

eellison commented:
We should copy over the no-op handling from the lowering because those are the only causes that actually show up in torchbench

Lezcano commented:
So, the benchmark above is not right. I didn't know I needed to register my decomposition in `torchinductor/decomposition.py`. As such, I was benchmarking the lowering against itself.
I'm adding the necessary stuff to benchmark the decomp now and I'll post the updated benchmark results.

jansel commented:
We should fix the bugs, so we are actually measuring the decomp.  Also, what hardware is this run on?

My instinct here is to stick with the lowering because it generates better code that will perform fewer loads using hardware masking registers.  If the perf is the same, we should have a preference towards the better code, because it may matter on different hardware, when this gets fused with other things, or other corner cases.

Lezcano commented:
For reference, the new benchmarks are here: https://github.com/pytorch/pytorch/pull/84483/
